### PR TITLE
[VRS-833] Make triplet loss an option in ultralytics library

### PR DESCRIPTION
**Pull request by sjhpark**
This modification implements triplet loss in YOLOv8 Pose model for contrastive learning.
It allows the user to create a docker container for running simple training with different task (e.g., detect, pose, pose-contrastive). Therefore, the user has options to train a YOLOv8 Pose model with either the default settings or contrastive learning (triplet loss & single label for classification).

-----------------------------TEST-----------------------------
1. Go to `~/verdant/dev/src/python/trainers/yolo8` in your server (e.g., neptune).
2. If you are in `master` branch of dev, the `Dockerfile` in this directory will clone a new verdant ultralytics repository's `main` branch. However, these changes (for triplet loss) have not been merged into the `main` branch yet. 
3. One work around is to checkout `jonghoon/yolo8_docker` branch in the `dev` repo. This branch has modified Dockerfile and `start.sh` to mount your local ultralytics repo to the docker container.
4. Thus, clone the verdant ultralytics repo locally and checkout to 'jonghoon/feature_embeddings' branch (this is the branch that I am trying to merge through this Pull Request).
Example: I cloned the verdant ultralytics repo as `ultralytics` in `~/verdant/dev/src/python/trainers/yolo8`.
6. Then, run  `./start.sh {your dataset path}/ runs {task} shell {local ultralytics dir}` from `~/verdant/dev/src/python/trainers/yolo8`. 
image
7. Example for **Pose** task: `./start.sh ~/data/onion4fira_24.9/ runs pose shell ~/verdant/dev/src/python/trainers/yolo8/ultralytics `
8. Example for **Pose-Contrastive** task: `./start.sh ~/data/onion4fira_24.9/ runs pose-contrastive shell ~/verdant/dev/src/python/trainers/yolo8/ultralytics `
9. Once the docker container is created and you are inside for interactive mode, you can type in `python simple_training.py` 
image
10. Once the training begins, you can monitor each loss and average logitis in the terminal. The triplet loss and average logits will always 0 for the tasks other than `Pose-Contrastive`. 
image

### CODE REVIEW

File: ultralytics/utils/loss.py [hunk 1]
```diff
@@ -524,6 +530,161 @@ def calculate_keypoints_loss(self, masks, target_gt_idx, keypoints, batch_idx, s
...
+            # for i in range(n_samples):
+            #   triplets = (embs_query[i], embs_pos[i], embs_neg[i])  # embeddings for query, positive, and negative samples
+            #   logit = torch.dot(triplets[0], (triplets[1] - triplets[2]))  # dot(query, pos) - dot(query, neg)
+            #   y = torch.ones_like(logit)  # (n_samples,)
+            #   triplet_loss += F.binary_cross_entropy_with_logits(logit, y)
+            diff = embs_pos - embs_neg  # (n_samples, n_features)
+            logits = torch.sum(embs_query * diff, dim=1)  # (n_samples,)
+            logits_avg += logits.mean().item()
+            y = torch.ones_like(logits)  # y is always 1 if we consider (query, pos) as a positive pair; (n_samples,)
+            triplet_loss += F.binary_cross_entropy_with_logits(logits, y, reduction='sum')
```
Comment (romainthibaux) [inline ultralytics/utils/loss.py:675]: / n_samples

File: ultralytics/nn/modules/head.py [hunk 2]
```diff
@@ -80,6 +79,102 @@ def bias_init(self):
...
+            return x
+        elif self.dynamic or self.shape != shape:
+            self.anchors, self.strides = (x.transpose(0, 1) for x in make_anchors(x, self.stride, 0.5))
+            self.shape = shape
+
+        x_cat = torch.cat([xi.view(shape[0], self.no, -1) for xi in x], 2)
+        if self.export and self.format in ('saved_model', 'pb', 'tflite', 'edgetpu', 'tfjs'):  # avoid TF FlexSplitV ops
+            box = x_cat[:, :self.reg_max * 4]
+            cls = x_cat[:, self.reg_max * 4: self.reg_max * 4 + self.nc]
+            _ = x_cat[:, self.reg_max * 4 + self.nc:]  # embeddings for contrastive learning
```
Comment (romainthibaux) [inline ultralytics/nn/modules/head.py:151]: Unused line

**Review by romainthibaux — CHANGES_REQUESTED** [review]

File: ultralytics/models/yolo/pose/train.py [hunk 3]
```diff
@@ -71,3 +71,36 @@ def plot_training_samples(self, batch, ni):
...
+class PoseContrastiveTrainer(PoseTrainer):
+    """
+    A class extending the DetectionTrainer class for training based on a pose model.
+
+    Example:
+        ```python
+        from ultralytics.models.yolo.pose import PoseTrainer
+
+        args = dict(model='yolov8n-pose.pt', data='coco8-pose.yaml', epochs=3)
+        trainer = PoseTrainer(overrides=args)
```
[the remark this answers is not in the document]
Comment (sjhpark) [reply ultralytics/models/yolo/pose/train.py:85]: Replaced `PoseTrainer` with `PoseContrastiveTrainer` in the docstring.

File: ultralytics/nn/tasks.py [hunk 4]
```diff
@@ -789,6 +800,8 @@ def cfg2task(cfg):
             return 'segment'
         if m == 'pose':
             return 'pose'
+        if m == 'posecontrastive':
```
[the remark this answers is not in the document]
Comment (sjhpark) [reply ultralytics/nn/tasks.py:803]: Mentioned `pose-contrastive` as a return value in the doc string.

File: ultralytics/utils/loss.py [hunk 5]
```diff
@@ -164,10 +164,10 @@ def bbox_decode(self, anchor_points, pred_dist):
...
         """Calculate the sum of the loss for box, cls and dfl multiplied by batch size."""
         loss = torch.zeros(3, device=self.device)  # box, cls, dfl
-        feats = preds[1] if isinstance(preds, tuple) else preds
-        pred_distri, pred_scores = torch.cat([xi.view(feats[0].shape[0], self.no, -1) for xi in feats], 2).split(
-            (self.reg_max * 4, self.nc), 1)
-
+        feats = preds[1] if isinstance(preds, tuple) else preds  # feats contains (predicted bboxes attributes, predicted classes, predcited embeddings)
+        pred_distri, pred_scores, embeddings = torch.cat([xi.view(feats[0].shape[0], self.no, -1) for xi in feats], 2).split(
+            (self.reg_max * 4, self.nc, self.no - (self.reg_max * 4 + self.nc)), 1)
+
```
[the remark this answers is not in the document]
Comment (sjhpark) [reply ultralytics/utils/loss.py:170]: Removed by reverting `v8DetectionLoss() `to the original.

File: ultralytics/utils/loss.py [hunk 6]
```diff
@@ -185,11 +185,11 @@ def __call__(self, preds, batch):
         # Pboxes
         pred_bboxes = self.bbox_decode(anchor_points, pred_distri)  # xyxy, (b, h*w, 4)
 
-        _, target_bboxes, target_scores, fg_mask, _ = self.assigner(
+        target_labels, target_bboxes, target_scores, fg_mask, _ = self.assigner(
             pred_scores.detach().sigmoid(), (pred_bboxes.detach() * stride_tensor).type(gt_bboxes.dtype),
             anchor_points * stride_tensor, gt_labels, gt_bboxes, mask_gt)
 
-        target_scores_sum = max(target_scores.sum(), 1)
+        target_scores_sum = max(target_scores.sum(), 1)  # target_scores.sum() is actually the same as fg_mask
```
[the remark this answers is not in the document]
Comment (sjhpark) [reply ultralytics/utils/loss.py:192]: Discarded by reverting `v8DetectionLoss() `to the original.

File: ultralytics/utils/loss.py [hunk 7]
```diff
@@ -205,6 +205,9 @@ def __call__(self, preds, batch):
         loss[1] *= self.hyp.cls  # cls gain
         loss[2] *= self.hyp.dfl  # dfl gain
 
+        # TODO: pull embeddings
+        # TODO: use target labels and embeddings to create triplet loss
+
```
[the remark this answers is not in the document]
Comment (sjhpark) [reply ultralytics/utils/loss.py:210]: Removed by reverting `v8DetectionLoss()` to the original.

File: ultralytics/utils/loss.py [hunk 8]
```diff
@@ -388,7 +391,7 @@ def __init__(self, model):  # model must be de-paralleled
 
     def __call__(self, preds, batch):
         """Calculate the total loss and detach it."""
-        loss = torch.zeros(5, device=self.device)  # box, cls, dfl, kpt_location, kpt_visibility
+        loss = torch.zeros(5+2, device=self.device)  # box, cls, dfl, kpt_location, kpt_visibility, placeholders for 0 triplet loss and 0 avg_logits
```
[the remark this answers is not in the document]
Comment (sjhpark) [reply ultralytics/utils/loss.py:394]: Replaces as suggested.

File: ultralytics/utils/loss.py [hunk 9]
```diff
@@ -442,6 +445,9 @@ def __call__(self, preds, batch):
         loss[3] *= self.hyp.cls  # cls gain
         loss[4] *= self.hyp.dfl  # dfl gain
 
+        # placeholder for 0 triplet loss and 0 avg_logits
+        loss[5], loss[6] = torch.tensor(0.0).to(dtype), torch.tensor(0.0).to(dtype)
+
```
[the remark this answers is not in the document]
Comment (sjhpark) [reply ultralytics/utils/loss.py:450]: Removed as suggested.

File: ultralytics/utils/loss.py [hunk 10]
```diff
@@ -524,6 +530,161 @@ def calculate_keypoints_loss(self, masks, target_gt_idx, keypoints, batch_idx, s
...
+class v8PoseContrastiveLoss(v8PoseLoss):
+    """Criterion class for computing training losses."""
+    def __init__(self, model):
+        """Initializes the v8PoseContrastiveLoss class."""
+        super().__init__(model)
+        self.n_samples = 10  # number of negative samples
+
+    def __call__(self, preds, batch):
+        """Calculate the total loss and detach it."""
+        loss = torch.zeros(5+2, device=self.device)  # box, cls, dfl, kpt_location, kpt_visibility, triplet loss, avg_logits
```
[the remark this answers is not in the document]
Comment (sjhpark) [reply ultralytics/utils/loss.py:542]: Replaces as suggested.

File: ultralytics/utils/loss.py [hunk 11]
```diff
@@ -524,6 +530,161 @@ def calculate_keypoints_loss(self, masks, target_gt_idx, keypoints, batch_idx, s
...
+        # # This is equivalent to Plant vs. None (background) classification.
+        # # This minimizes classification’s impact on generating embeddings 
+        # # while triplet loss is encouraging the model to distinguish one class from another.
+        label_for_single_class_cls = 0
+        gt_labels_single_class_cls = torch.full_like(gt_labels, label_for_single_class_cls)
+        _, target_bboxes, target_scores, fg_mask, target_gt_idx = self.assigner(
+            pred_scores.detach().sigmoid(), (pred_bboxes.detach() * stride_tensor).type(gt_bboxes.dtype),
+            anchor_points * stride_tensor, gt_labels_single_class_cls, gt_bboxes, mask_gt)
+
+        target_scores_sum = max(target_scores.sum(), 1)
```
[the remark this answers is not in the document]
Comment (sjhpark) [reply ultralytics/utils/loss.py:582]: We revisited and discussed, and concluded that `target_scores_sum` is not always 1, but it is rather the sum of ground truth labels having matches and this number can change by batch.

File: ultralytics/nn/modules/head.py (hunk 2 above)
Comment (sjhpark) [reply ultralytics/nn/modules/head.py:151]: Removed.

File: ultralytics/nn/modules/head.py [hunk 12]
```diff
@@ -80,6 +79,101 @@ def bias_init(self):
...
+        """Initializes the YOLOv8 detection layer with specified number of classes and channels.
+        This implementation has an additional contrastive learning task for embeddings of predicted bboxes.
+        This expands (doubles) the out_channels of the last convolution layer before classification layer 
+        to split embedding dimension for classification and contrastive learning tasks.
+        """
+        super().__init__()
+        self.nc = nc  # number of classes
+        self.nl = len(ch)  # number of detection layers (e.g., small, medium, large for 3-scale detections)
+        self.reg_max = 16  # DFL channels (ch[0] // 16 to scale 4/8/12/16/20 for n/s/m/l/x)
+        c2, c3 = max((16, ch[0] // 4, self.reg_max * 4)), max(ch[0], min(self.nc, 100))  # channels
```
Comment (romainthibaux) [inline ultralytics/nn/modules/head.py:100]: Extraneous parentheses.

File: ultralytics/nn/modules/head.py [hunk 13]
```diff
@@ -80,6 +79,101 @@ def bias_init(self):
...
+        self.no = nc + self.reg_max * 4 + c3  # number of outputs per anchor (nc is num classes, reg_max is # predicted bboxes & 4 is bbox attributes (x, y, w, h), c3 is num features in final classification layer's embeddings)
+        self.stride = torch.zeros(self.nl)  # strides computed during build
+
+        # cv2 is for predicting bboxes (last layer's out_channel is 4 * reg_max)
+        self.cv2 = nn.ModuleList(
+            nn.Sequential(Conv(x, c2, 3), Conv(c2, c2, 3), nn.Conv2d(c2, 4 * self.reg_max, 1)) for x in ch)
+        
+        # expand out_channels for conv layer before classification layer to split embedding dimension for classification and contrastive learning tasks
+        self.c3_cls_contrastive = c3 * 2  # e.g., 64 + 64 = 128
```
Comment (romainthibaux) [inline ultralytics/nn/modules/head.py:109]: It would be clearer and more flexible if `self.c3_cls_contrastive` represented the number of channels for the embeddings, rather than the sum of classification feature vector plus embedding. For instance, right now you have no easy way to express the case where the number of channels for these two purposes is different (if you wanted to). So I would set
```
self.c3_cls_contrastive = c3
```
and later use
```
Conv(c3, self.c3 + self.c3_cls_contrastive, 3)
```
Then in `forward()` you can write more simply:
```
            embeddings_cls = self.embedding_layers[i](x[i])[:, :self.c3, :, :]  # (B, first half of n_features, H, W)
            embeddings_contrastive = self.embedding_layers[i](x[i])[:, self.c3:, :, :]  # (B, second half of n_features, H, W)
```

File: ultralytics/nn/modules/head.py [hunk 14]
```diff
@@ -80,6 +79,101 @@ def bias_init(self):
...
+        This implementation has an additional contrastive learning task for embeddings of predicted bboxes.
+        This expands (doubles) the out_channels of the last convolution layer before classification layer 
+        to split embedding dimension for classification and contrastive learning tasks.
+        """
+        super().__init__()
+        self.nc = nc  # number of classes
+        self.nl = len(ch)  # number of detection layers (e.g., small, medium, large for 3-scale detections)
+        self.reg_max = 16  # DFL channels (ch[0] // 16 to scale 4/8/12/16/20 for n/s/m/l/x)
+        c2, c3 = max((16, ch[0] // 4, self.reg_max * 4)), max(ch[0], min(self.nc, 100))  # channels
+        self.no = nc + self.reg_max * 4 + c3  # number of outputs per anchor (nc is num classes, reg_max is # predicted bboxes & 4 is bbox attributes (x, y, w, h), c3 is num features in final classification layer's embeddings)
```
Comment (romainthibaux) [inline ultralytics/nn/modules/head.py:101]: If `self.no` represents the number of output channels per anchor, then this is wrong. It should include the embedding channels.
The only place where you are using this value is in the non-training branch of `forward()`, which is probably untested. I don't think we need to worry too much about testing that portion of the code. I assumed it was used when exporting to onnx but if you are able to export your models then this code may not actually be used. Nonetheless, we should try to keep it correct, and I think right now it would fail on line 147:
```
x_cat = torch.cat([xi.view(shape[0], self.no, -1) for xi in x], 2)
```
because the dimension `self.no` does not match that of the tensors `xi`.
Do you know whether that code is used?

**Review by romainthibaux — CHANGES_REQUESTED** [review]

File: ultralytics/nn/modules/head.py (hunk 12 above)
Comment (sjhpark) [reply ultralytics/nn/modules/head.py:100]: Removed the extra parentheses.
These extra parentheses came from the original Ultralytics code, so I removed these from `Detect()` class as well.
`c2, c3 = max((16, ch[0] // 4, self.reg_max * 4)), max(ch[0], min(self.nc, 100) --> c2, c3 = max(16, ch[0] // 4, self.reg_max * 4), max(ch[0], min(self.nc, 100)`

File: ultralytics/nn/modules/head.py (hunk 14 above)
Comment (sjhpark) [reply ultralytics/nn/modules/head.py:101]: Correct. For PoseContrastive task, self.no (# output channels per anchor) should use the # features of embedding for contrastive learning, not classification.
I revised as below to have `self.no` use `self.c3_cls_contrastive` rather than `self.c3`.
```
        # define the number of features in embedding for contrastive learning task
        self.c3_cls_contrastive = self.c3

        # number of outputs per anchor
        # (nc is # classes, reg_max is # predicted bboxes & 4 is bbox attributes (x, y, w, h), c3_cls_contrastive is # features in final classification layer's embeddings for contrastive learning)
        self.no = nc + self.reg_max * 4 + self.c3_cls_contrastive
```

------------------
**Findings from running simple_training.py (this is for documenting my observations here so that I can refer back later):**
- When we run `simple_training.py`, the code runs inference twice and runs training several times before starting the actual training.
- Each inference run and training run **seem to be a test run** (not confirmed, but it looks like to me).
- The very first inference and training runs use 1 for `nc (num classes)`. The next runs seem to use the dataset's num classes.
- Each inference uses 3 samples as input dataset and `8, 4, and 1 as (large, med, or small)` as detection scales. Maybe it is mimicking 96, 48, 24 detection scales?
- Each training uses 3 samples as input dataset and `32, 16, and 8 as (large, med, or small)` as detection scales.
- Once the actual training begins, it uses `96, 48, and 24 as (large, med, or small)` as detection scales.

Example:
```
# nc: # classes
# reg_max: # predicted bboxes
# 4: bbox attributes

# first test run:
no = 1 + self.reg_max * 4 + self.c3_cls_contrastive  # number of output channels per anchor (detection candidate)
torch.Size([1, no, 32, 32]) torch.Size([1, no, 16, 16]) torch.Size([1, no, 8, 8])  # training
torch.Size([1, no, 4, 4]) torch.Size([1, no, 2, 2]) torch.Size([1, no, 1, 1])  # inference

# second and so forth:
no = nc + self.reg_max * 4 + self.c3_cls_contrastive  # number of output channels per anchor (detection candidate)
torch.Size([1, no, 32, 32]) torch.Size([1, no, 16, 16]) torch.Size([1, no, 8, 8])  # training
torch.Size([1, no, 4, 4]) torch.Size([1, no, 2, 2]) torch.Size([1, no, 1, 1])  # inference
```

**Review by romainthibaux — APPROVED** [review]
